ProjFS.Mac/PrjFSKext/VnodeCache.cpp

```diff
@@ -0,0 +1,414 @@
+}
+
+void VnodeCache_InvalidateCache(PerfTracer* _Nonnull perfTracer)
+{
+    perfTracer->IncrementCount(PrjFSPerfCounter_CacheInvalidateCount, true /*ignoreSampling*/);
+
+    RWLock_AcquireExclusive(s_entriesLock);
+    {
+        InvalidateCache_ExclusiveLocked();
+    }
+    RWLock_ReleaseExclusive(s_entriesLock);
+}
+
+KEXT_STATIC_INLINE void InvalidateCache_ExclusiveLocked()
+{
+    memset(s_entries, 0, s_entriesCapacity * sizeof(VnodeCacheEntry));
+}
+
+KEXT_STATIC_INLINE uint32_t ComputePow2CacheCapacity(int expectedVnodeCount)
+{
+    uint32_t idealCacheSize = expectedVnodeCount * 2;
+    uint32_t capacity = MinPow2VnodeCacheCapacity;
+    while (capacity < idealCacheSize && capacity < MaxPow2VnodeCacheCapacity)
+    {
+        capacity = capacity << 1;
+    }
+   
+    return capacity;
+}
+
+KEXT_STATIC_INLINE uintptr_t ComputeVnodeHashIndex(vnode_t _Nonnull vnode)
+{
+    uintptr_t vnodeAddress = reinterpret_cast<uintptr_t>(vnode);
+    return (vnodeAddress >> 3) & s_ModBitmask;
+}
+
+KEXT_STATIC bool TryGetVnodeRootFromCache(
+    vnode_t _Nonnull vnode,
+    uintptr_t vnodeHashIndex,
+    uint32_t vnodeVid,
+    /* out parameters */
+    VirtualizationRootHandle& rootHandle)
+{
+    bool rootFound = false;
+    rootHandle = RootHandle_None;
+
+    RWLock_AcquireShared(s_entriesLock);
+    {
+        uintptr_t vnodeIndex;
+        if (TryFindVnodeIndex_Locked(vnode, vnodeHashIndex, /*out*/ vnodeIndex))
+        {
+            if (vnode == s_entries[vnodeIndex].vnode &&
+                vnodeVid == s_entries[vnodeIndex].vid &&
+                RootHandle_Indeterminate != s_entries[vnodeIndex].virtualizationRoot)
+            {
+                rootFound = true;
+                rootHandle = s_entries[vnodeIndex].virtualizationRoot;
+            }
+        }
+    }
+    RWLock_ReleaseShared(s_entriesLock);
+    
+    return rootFound;
+}
+
+KEXT_STATIC void FindVnodeRootFromDiskAndUpdateCache(
+    PerfTracer* _Nonnull perfTracer,
+    PrjFSPerfCounter cacheMissFallbackFunctionCounter,
+    PrjFSPerfCounter cacheMissFallbackFunctionInnerLoopCounter,
+    vfs_context_t _Nonnull context,
+    vnode_t _Nonnull vnode,
+    uintptr_t vnodeHashIndex,
+    uint32_t vnodeVid,
+    UpdateCacheBehavior updateEntryBehavior,
+    /* out parameters */
+    VirtualizationRootHandle& rootHandle)
+{
+    rootHandle = VirtualizationRoot_FindForVnode(
+        perfTracer,
+        cacheMissFallbackFunctionCounter,
+        cacheMissFallbackFunctionInnerLoopCounter,
+        vnode,
+        context);
+
+    bool forceRefreshEntry;
+    VirtualizationRootHandle rootToInsert;
+    switch (updateEntryBehavior)
+    {
+      case UpdateCacheBehavior_ForceRefresh:
+        rootToInsert = rootHandle;
+        forceRefreshEntry = true;
+        break;
+        
+      case UpdateCacheBehavior_InvalidateEntry:
+        rootToInsert = RootHandle_Indeterminate;
+        forceRefreshEntry = true;
+        break;
+
+      default:
+        rootToInsert = rootHandle;
```

Comment:
I'm fine with the way this is structured.

I've also see the approach of 

UpdateCacheBehavior_TrustCurrentEntry:

default:
  throw exception unknown type

I think what we have here makes sense for  the kext

Reply:
I will add a real entry for `UpdateCacheBehavior_TrustCurrentEntry` and an `assertf` in the default case.